plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb

```diff
@@ -3307,15 +3307,26 @@ def self.select_best_fit_node_impl(server_infos)
       end
       server_infos.delete_if { |server_info| !server_info.zone_id } if has_zone_node
 
-      # Sort by node active capacity (consumed capacity) and take the best half
-      server_infos = server_infos.sort_by { |server_info| server_info.node_consumed_capacity }
-      # consider the top half and no less than min(4, the actual number of available)
-      server_infos = server_infos.first([4, (server_infos.length / 2).to_i].max)
-
-      # Sort by district available capacity and take the best half
-      server_infos = server_infos.sort_by { |server_info| (server_info.district_available_capacity ? server_info.district_available_capacity : 1) }
-      # consider the top half and no less than min(4, the actual number of available)
-      server_infos = server_infos.last([4, (server_infos.length / 2).to_i].max)
+      if server_infos.length > 4
+        # Take all nodes with < the average capacity.  Make sure to take at least max(4, 20% of nodes) nodes.
+        server_infos = server_infos.sort_by { |server_info| server_info.node_consumed_capacity }.reverse!
+        node_consumed_capacities = server_infos.map { |server_info| server_info.node_consumed_capacity }
+        average_consumed_capacity = node_consumed_capacities.inject(0.0) { |sum, c| sum + c } / node_consumed_capacities.length
+        min_nodes = [4, (server_infos.length * 0.2).to_i].max
+        server_infos.delete_if { |server_info| server_info.node_consumed_capacity > average_consumed_capacity && server_infos.length > min_nodes }
```

Comment:
The intent here is a little unclear, but delete_if doesn't actually delete anything (have to put the result back in the var) and server_infos.length wouldn't change during the loop. If the idea is to chop out everything that's above average, but not below min, then it would probably be best to sort and have the loop compare the index to the min_nodes to make sure you don't start chopping too soon.

Test case might help...